lib/install/phases/coredns.go

```diff
@@ -0,0 +1,176 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package phases
+
+import (
+	"bytes"
+	"context"
+
+	"github.com/alecthomas/template"
+	"github.com/gravitational/gravity/lib/constants"
+	"github.com/gravitational/gravity/lib/defaults"
+	"github.com/gravitational/gravity/lib/fsm"
+	"github.com/gravitational/gravity/lib/ops"
+	"github.com/gravitational/gravity/lib/storage"
+	"github.com/gravitational/gravity/lib/systeminfo"
+	"github.com/gravitational/trace"
+	log "github.com/sirupsen/logrus"
+	"k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+// corednsExecutor is the phase which generated CoreDNS configuration for the cluster
+type corednsExecutor struct {
+	// FieldLogger specifies the logger used by the executor
+	log.FieldLogger
+	// ExecutorParams contains common executor parameters
+	fsm.ExecutorParams
+	// Client is the Kubernetes client
+	Client *kubernetes.Clientset
+	// DNSOverrides is the user configured DNS overrides
+	DNSOverrides storage.DNSOverrides
+}
+
+// NewCorednsPhase creates a new coredns phase executor
+func NewCorednsPhase(p fsm.ExecutorParams, operator ops.Operator, client *kubernetes.Clientset) (fsm.PhaseExecutor, error) {
+	logger := &fsm.Logger{
+		FieldLogger: log.WithField(constants.FieldPhase, p.Phase.ID),
+		Key:         opKey(p.Plan),
+		Operator:    operator,
+	}
+
+	cluster, err := operator.GetSite(ops.SiteKey{
+		AccountID:  defaults.SystemAccountID,
+		SiteDomain: p.Plan.ClusterName,
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &corednsExecutor{
+		FieldLogger:    logger,
+		ExecutorParams: p,
+		Client:         client,
+		DNSOverrides:   cluster.DNSOverrides,
+	}, nil
+}
+
+// PreCheck is no-op for this phase
+func (r *corednsExecutor) PreCheck(context.Context) error {
+	return nil
+}
+
+// PostCheck is no-op for this phase
+func (r *corednsExecutor) PostCheck(context.Context) error {
+	return nil
+}
+
+// Execute generated coredns configuration
+func (r *corednsExecutor) Execute(ctx context.Context) error {
+	r.Progress.NextStep("Configuring Coredns")
+	r.Info("Configuring Coredns.")
+
+	// Read the resolv.conf from the host doing installation
+	// it will be used for configuring coredns upstream servers
+	resolvConf, err := systeminfo.ResolvFromFile("/etc/resolv.conf")
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	conf, err := GenerateCorefile(CorednsConfig{
+		UpstreamNameservers: resolvConf.Servers,
+		Rotate:              resolvConf.Rotate,
+		Hosts:               r.DNSOverrides.Hosts,
+		Zones:               r.DNSOverrides.Zones,
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	_, err = r.Client.CoreV1().ConfigMaps(constants.KubeSystemNamespace).Create(&v1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "coredns",
```

Comment:
Should the name of the CoreDNS ConfigMap also move to constants/defaults? It is being used in multiple places as far as I can see.